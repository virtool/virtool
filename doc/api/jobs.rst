Jobs
====


Find
----

Retrieve a list of job documents based on filters defined in URL parameters.

::

    GET /api/jobs?term=nuvs&sort=timestamp


**Parameters**

+----------+---------+----------------------------------+
| Name     | Type    | Description                      |
+==========+=========+==================================+
| term     | string  | filters by username or task name |
+----------+---------+----------------------------------+
| sort     | string  | a document field to sort by      |
+----------+---------+----------------------------------+

**Response**

.. code-block:: javascript

    [
        {
            "job_id": "f41e8c"
            "task": "add_host",
            "proc": 1,
            "mem": 4,
            "state": "running",
            "stage": "snap_build",
            "progress": 0.571,
            "added": "2016-08-25T11:48:22.375557"
            "user_id": "igboyes"
        },
        {
            "job_id": "jk98ty"
            "task": "import_reads",
            "proc": 1,
            "mem": 4,
            "state": "running",
            "stage": "snap_build",
            "progress": 0.231,
            "added": "2016-08-25T11:48:22.375557"
            "user_id": "igboyes"
        },
        {
            "job_id": "89thy0"
            "task": "pathoscope_bowtie",
            "proc": 1,
            "mem": 4,
            "state": "running",
            "stage": "map_host",
            "progress": 1.0,
            "added": "2016-04-25T11:23:11.375557"
            "user_id": "igboyes"
        }
    ]



Get
---

Retrieve a complete job document from the server.

::

    GET /api/jobs/job_id


**Response**

.. code-block:: javascript

    {
        "job_id": "f41e8c"
        "task": "add_host",
        "args": {
            "_id": "Prunus persica",
            "_version": 0,
            "description": "Peach",
            "file": "prunus_persica.fa",
            "added": false,
            "job": "f41e8c",
            "username": "igboyes"
        },
        "proc": 1,
        "mem": 4,
        "status": [
            {
                "stage" : "bowtie_build",
                "progress" : 0.429,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:48:22.375557"
            },
            {
                "stage" : "snap_build",
                "progress" : 0.571,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:52:18.472402"
            },
            {
                "stage" : "update_db",
                "progress" : 0.714,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:53:24.774645"
            },
        ],
        "user_id": "igboyes"
    }



Cancel
------

Cancel a running or waiting job.

::

    PUT /api/jobs/:job_id


**Input**

+----------+---------+----------------------------------+
| Name     | Type    | Description                      |
+==========+=========+==================================+
| cancel   | bool    | **must** be ``true``             |
+----------+---------+----------------------------------+

**Response**

.. code-block:: javascript

    {
        "job_id": "f41e8c"
        "task": "add_host",
        "args": {
            "_id": "Prunus persica",
            "_version": 0,
            "description": "Peach",
            "file": "prunus_persica.fa",
            "added": false,
            "job": "f41e8c",
            "username": "igboyes"
        },
        "proc": 1,
        "mem": 4,
        "status": [
            {
                "stage" : "bowtie_build",
                "progress" : 0.429,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:48:22.375557"
            },
            {
                "stage" : "snap_build",
                "progress" : 0.571,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:52:18.472402"
            },
            {
                "stage" : "update_db",
                "progress" : 0.714,
                "error" : null,
                "state" : "running",
                "date" : "2016-08-25T11:53:24.774645"
            },
            {
                "stage" : "update_db",
                "progress" : 0.714,
                "error" : null,
                "state" : "cancelled",
                "date" : "2016-08-25T11:55:54.475545"
            },
        ],
        "user_id": "igboyes"
    }



Remove
------

Remove a job document. The job must not be running or waiting

::

    DELETE /api/jobs/:job_id

**Response**

.. code-block:: javascript

    {
        "removed": "f41e8c"
    }
