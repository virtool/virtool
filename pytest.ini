[pytest]
addopts = --aiohttp-loop uvloop --capture tee-sys
asyncio_mode = auto
