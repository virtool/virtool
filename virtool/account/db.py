"""
Work with the current user account and its API keys.

API Key Schema:
- _id (str) the hashed API key - never returned
- id (str) the API key ID
- name (str) user-defined name for the API key
- groups (Array[str]) list of groups the API key user is a member of
- permissions (Object) user-defined permissions possessed by the key - permission names are keys with boolean values
- created_at (datetime) when the API key was created)
- user (Object) describes the parent user
    id (str) the user ID

"""
import virtool.account.utils
import virtool.users.db
import virtool.users.utils
import virtool.utils

PROJECTION = [
    "_id",
    "administrator",
    "email",
    "groups",
    "identicon",
    "last_password_change",
    "permissions",
    "primary_group",
    "settings"
]


def compose_password_update(password: str) -> dict:
    """
    Compose an update dict for self-changing a users account password. This will disable forced reset and won't
    invalidate current sessions, unlike a password change by an administrator.

    :param password: the new password
    :return: a password update

    """
    # Update the user document. Remove all sessions so those clients will have to authenticate with the new
    # password.
    return {
        "password": virtool.users.utils.hash_password(password),
        "invalidate_sessions": False,
        "last_password_change": virtool.utils.timestamp(),
        "force_reset": False
    }


async def get(db, user_id: str) -> dict:
    """
    Get appropriately projected user document by id.

    :param db: the application database object
    :param user_id: the user id
    :return: the projected user document

    """
    return await db.users.find_one(
        user_id,
        PROJECTION
    )


async def get_alternate_id(db, name: str) -> str:
    """
    Get an alternate id for an API key whose provided `name` is not unique. Appends an integer suffix to the end of the
    `name`.

    :param db: the application database object
    :param name: the API key name
    :return: an alternate unique id for the key

    """
    existing_alt_ids = await db.keys.distinct("id")

    suffix = 0

    while True:
        candidate = f"{name.lower()}_{suffix}"

        if candidate not in existing_alt_ids:
            return candidate

        suffix += 1


async def create_api_key(db, name: str, permissions: dict, user_id: str) -> dict:
    """
    Create a new API key for the account with the given `user_id`.

    API keys can only receive permissions possessed by the owner of the API key. If the owner is an administrator, their
    key permissions will not be limited.

    Actions that require administrator status cannot be performed using API key authentication.

    :param db: the application database object
    :param name: a display name for the API key
    :param permissions: permissions to provide to the API key
    :param user_id: the id of the owning user
    :return: the API key document

    """
    user = await db.users.find_one(user_id, ["administrator", "groups", "permissions"])

    key_permissions = {
        **virtool.users.utils.generate_base_permissions(),
        **permissions
    }

    if not user["administrator"]:
        key_permissions = virtool.users.utils.limit_permissions(key_permissions, user["permissions"])

    raw, hashed = virtool.account.utils.generate_api_key()

    document = {
        "_id": hashed,
        "id": await virtool.account.db.get_alternate_id(db, name),
        "name": name,
        "groups": user["groups"],
        "permissions": key_permissions,
        "created_at": virtool.utils.timestamp(),
        "user": {
            "id": user_id
        }
    }

    await db.keys.insert_one(document)

    del document["_id"]
    del document["user"]

    document["key"] = raw

    return document
