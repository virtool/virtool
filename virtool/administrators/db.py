from typing import TYPE_CHECKING

from motor.motor_asyncio import AsyncIOMotorClientSession
from virtool_core.models.roles import AdministratorRole

if TYPE_CHECKING:
    from virtool.mongo.core import Mongo


async def update_legacy_administrator(
    mongo: "Mongo",
    user_id: str,
    role: AdministratorRole,
    session: AsyncIOMotorClientSession = None,
):
    """
    Update the legacy administrator field for a user.
    """

    await mongo.users.update_one(
        {"_id": user_id},
        {"$set": {"administrator": role == AdministratorRole.FULL}},
        session=session,
    )
