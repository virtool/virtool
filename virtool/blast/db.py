from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from virtool.blast.models import SQLNuVsBlast


async def get_nuvs_blast(
    session: AsyncSession,
    analysis_id: str,
    sequence_index: int,
) -> SQLNuVsBlast:
    result = await session.scalar(
        select(SQLNuVsBlast)
        .where(SQLNuVsBlast.analysis_id == analysis_id)
        .where(SQLNuVsBlast.sequence_index == sequence_index),
    )

    return result


async def delete_nuvs_blast(
    session: AsyncSession,
    analysis_id: str,
    sequence_index: int,
) -> int:
    result = await session.execute(
        delete(SQLNuVsBlast)
        .where(SQLNuVsBlast.analysis_id == analysis_id)
        .where(SQLNuVsBlast.sequence_index == sequence_index),
    )

    await session.commit()

    return result.rowcount
