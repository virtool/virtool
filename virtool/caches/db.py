"""
Work with caches in the database. Caches are bundles of trimmed read and QC data generated during analyses.

Schema:
- _id (str) the ID of the cache
- created_at (datetime) when the cache record was created
- files (Array[Object]) describes trimmed FASTQ files
  - name (str) the name of the trimmed read file on disk
  - size (int) the size of the file
- hash (str) a hash that uniquely identifies the trim - calculated from the trim parameters used to create the cache
- missing (bool) set to true if the cache cannot be found on disk
- paired (bool) set to true if the cache represents paired data
- parameters (Object) the parameters used to trim the data
- program (str) the name of the program used to trim the data (eg. skewer-0.2.2)
- quality (JSON) the FastQC output converted to a JSON object
- ready (bool) set to true when the cache has be successfully created
- sample (Object) describes the sample the cache is dervied from
  - id (str) the sample ID

"""
import asyncio
import hashlib
import json
import os
from typing import Optional

import aiohttp.web
import pymongo.errors

import virtool.caches
import virtool.utils

PROJECTION = [
    "_id",
    "created_at",
    "files",
    "hash",
    "program",
    "ready",
    "sample"
]


def calculate_cache_hash(parameters: dict) -> str:
    """
    Calculate a hash from the parameters `dict` for a cache. The parameters are arguments passed to a trimming program.
    Caches can be reused when the hash of the trim parameters for a a new analysis matches an existing cache.

    :param parameters: the trimming parameters
    :return: the cache hash

    """
    string = json.dumps(parameters, sort_keys=True)
    return hashlib.sha1(string.encode()).hexdigest()


async def find(db, sample_id: str, program: str, parameters: dict) -> Optional[dict]:
    """
    Find a cache matching the passed `sample_id`, `program` name and version, and set of trimming `parameters`.

    If no matching cache exists, `None` will be returned.

    :param db: the application database interface
    :param sample_id: the id of the parent sample
    :param program: the program and version used to create the cache
    :param parameters: the parameters used for the trim
    :return: a cache document

    """
    document = await db.caches.find_one({
        "hash": virtool.caches.db.calculate_cache_hash(parameters),
        "missing": False,
        "program": program,
        "sample.id": sample_id
    })

    return virtool.utils.base_processor(document)


async def find_and_wait(db, sample_id: str, program: str, parameters: dict) -> Optional[dict]:
    """
    Find a cache matching the passed `sample_id`, `program` name and version, and set of trimming `parameters`. Wait
    for the cache to be ready if it is still being created.

    If no matching cache exists, `None` will be returned.

    :param db: the application database interface
    :param sample_id: the id of the parent sample
    :param program: the program and version used to create the cache
    :param parameters: the parameters used for the trim
    :return: a cache document

    """
    document = await find(db, sample_id, program, parameters)

    if document:
        cache_id = document["id"]

        while document["ready"] is False:
            await asyncio.sleep(2)
            document = virtool.utils.base_processor(await db.caches.find_one(cache_id))

    return virtool.utils.base_processor(document)


async def get(db, cache_id: str) -> dict:
    """
    Get the complete representation for the cache with the given `cache_id`.

    :param db: the application database client
    :param cache_id: the id of the cache to get
    :return: the cache document

    """
    document = await db.caches.find_one(cache_id)
    return virtool.utils.base_processor(document)


async def create(db, sample_id: str, parameters: dict, paired: bool, legacy: bool = False, program: str = "skewer-0.2.2"):
    """
    Create and insert a new cache database document. Return the generated unique cache id.

    :param db: the application database client
    :param sample_id: the id of the sample the cache is derived from
    :param parameters: the trim parameters
    :param paired: boolean indicating if the sample contains paired data
    :param legacy: boolean indicating if the cache is derived from a trimmed legacy sample
    :param program: the trimming program used
    :return: the new cache id

    """
    try:
        cache_id = virtool.utils.random_alphanumeric(length=8)

        document = {
            "_id": cache_id,
            "created_at": virtool.utils.timestamp(),
            "files": list(),
            "hash": calculate_cache_hash(parameters),
            "legacy": legacy,
            "missing": False,
            "paired": paired,
            "parameters": parameters,
            "program": program,
            "ready": False,
            "sample": {
                "id": sample_id
            }
        }

        await db.caches.insert_one(document)

        return virtool.utils.base_processor(document)

    except pymongo.errors.DuplicateKeyError:
        # Keep trying to add the cache with new ids if the generated id is not unique.
        return await create(db, sample_id, parameters, paired, legacy=legacy, program=program)


async def remove(app: aiohttp.web.Application, cache_id: str):
    """
    Remove the cache database document and files with the given `cache_id`.

    :param app: the application object
    :param cache_id: the id of the cache to remove

    """
    db = app["db"]
    settings = app["settings"]

    await db.caches.delete_one({
        "_id": cache_id
    })

    path = os.path.join(settings["data_path"], "caches", cache_id)

    try:
        await app["run_in_thread"](virtool.utils.rm, path, True)
    except FileNotFoundError:
        pass
