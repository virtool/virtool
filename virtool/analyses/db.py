"""
Work with analyses in the database.

Schema:
- _id (str) the ID for the analysis
- cache (Object) describes the cache associated with the sample
  - id (str) the ID of the cache
- created_at (datetime) when the analyses document was first created
- index (Object) describes the index analyzed against
  - id (str) the index ID
- job (Object) describes the workflow job
  - id (str) the job ID
- read_count (int)
  - the number of reads involved in the analysis
- ready (bool) true when the analysis workflow has completed
- reference (Object) described the reference analyzed against
  - id (str) the ID of the reference
- results (JSON) the result payload of the workflow
- sample (Object) describes the sample analyzed
  - id (str) the sample ID
- subtracted_count (int) number of reads subtracted (should be in results)
- subtraction (Object) describes the subtraction
  - id (str) the subtraction ID
- updated_at (datetime) when the analysis was last modified - used for browser caching
- user (Object) describes the creating user
  - id (str) the user ID
- workflow (Enum["aodp", "nuvs", "pathoscope") the workflow

"""
import asyncio
import os
from typing import Tuple, Optional

import virtool.analyses.utils
import virtool.bio
import virtool.db.utils
import virtool.history.db
import virtool.indexes.db
import virtool.jobs.db
import virtool.otus.utils
import virtool.samples.db
import virtool.types
import virtool.utils

PROJECTION = [
    "_id",
    "workflow",
    "created_at",
    "index",
    "job",
    "ready",
    "reference",
    "sample",
    "subtraction",
    "updated_at",
    "user"
]


class BLAST:

    def __init__(self, db, settings: dict, analysis_id: str, sequence_index: int, rid: str):
        self.db = db
        self.settings = settings
        self.analysis_id = analysis_id
        self.sequence_index = sequence_index
        self.rid = rid
        self.error = None
        self.interval = 3
        self.ready = False
        self.result = None

    async def remove(self):
        """
        Remove the BLAST result from the analysis document.

        """
        await remove_nuvs_blast(self.db, self.analysis_id, self.sequence_index)

    async def sleep(self):
        """
        Sleep for the current interval and increase the interval by 5 seconds after sleeping.

        """
        await asyncio.sleep(self.interval)
        self.interval += 5

    async def update(self, ready: bool, result: Optional[dict], error: Optional[str]) -> Tuple[dict, dict]:
        """
        Update the BLAST data. Returns the BLAST data and the complete analysis document.

        :param ready: indicates whether the BLAST request is complete
        :param result: the formatted result of a successful BLAST request
        :param error: and error message to add to the BLAST record
        :return: the BLAST data and the complete analysis document

        """
        self.result = result

        if ready is None:
            self.ready = await virtool.bio.check_rid(self.settings, self.rid)
        else:
            self.ready = ready

        data = {
            "error": error,
            "interval": self.interval,
            "last_checked_at": virtool.utils.timestamp(),
            "rid": self.rid,
            "ready": ready,
            "result": self.result
        }

        document = await self.db.analyses.find_one_and_update({
            "_id": self.analysis_id,
            "results.index": self.sequence_index
        }, {
            "$set": {
                "results.$.blast": data,
                "updated_at": virtool.utils.timestamp()
            }
        })

        return data, document


async def create(app: virtool.types.App, sample_id: str, ref_id: str, subtraction_id: str, user_id: str, workflow: str) -> dict:
    """
    Creates a new analysis. Ensures that a valid subtraction host was the submitted. Configures read and write
    permissions on the sample document and assigns it a creator username based on the requesting connection.

    :param app: the application object
    :param sample_id: the ID of the sample to create an analysis for
    :param ref_id: the ID of the reference to analyze against
    :param subtraction_id: the ID of the subtraction to remove from the analysis
    :param user_id: the ID of the user starting the job
    :param workflow: the analysis workflow to run
    :return: the analysis document

    """
    db = app["db"]
    settings = app["settings"]

    # Get the current id and version of the otu index currently being used for analysis.
    index_id, index_version = await virtool.indexes.db.get_current_id_and_version(db, ref_id)

    sample = await db.samples.find_one(sample_id, ["name"])

    analysis_id = await virtool.db.utils.get_new_id(db.analyses)

    job_id = await virtool.db.utils.get_new_id(db.jobs)

    created_at = virtool.utils.timestamp()

    document = {
        "_id": analysis_id,
        "ready": False,
        "created_at": created_at,
        "updated_at": created_at,
        "job": {
            "id": job_id
        },
        "workflow": workflow,
        "sample": {
            "id": sample_id
        },
        "index": {
            "id": index_id,
            "version": index_version
        },
        "reference": {
            "id": ref_id,
            "name": await virtool.db.utils.get_one_field(db.references, "name", ref_id)
        },
        "subtraction": {
            "id": subtraction_id
        },
        "user": {
            "id": user_id,
        }
    }

    task_args = {
        "analysis_id": analysis_id,
        "ref_id": ref_id,
        "sample_id": sample_id,
        "sample_name": sample["name"],
        "index_id": index_id
    }

    await db.analyses.insert_one(document)

    # Create job document.
    job = await virtool.jobs.db.create(
        db,
        document["workflow"],
        task_args,
        user_id
    )

    await app["jobs"].enqueue(job["_id"])

    await virtool.samples.db.recalculate_workflow_tags(db, sample_id)

    return document


async def update_nuvs_blast(
        db,
        settings: dict,
        analysis_id: str,
        sequence_index: int,
        rid: str,
        error: Optional[str] = None,
        interval: int = 3,
        ready: Optional[bool] = None,
        result: Optional[dict] = None
) -> Tuple[dict, dict]:
    """
    Update the BLAST data for a sequence in a NuVs analysis.

    :param db: the application database object
    :param settings: the application settings
    :param analysis_id: the id of the analysis the BLAST is for
    :param sequence_index: the index of the NuVs sequence the BLAST is for
    :param rid: the id of the request
    :param error: an error message if the BLAST failed
    :param interval: the current interval for checking the request status on NCBI
    :param ready: indicates that the BLAST result is ready
    :param result: the formatted result from NCBI
    :return: the blast data and the complete analysis document

    """
    if ready is None:
        ready = await virtool.bio.check_rid(settings, rid)

    data = {
        "interval": interval,
        "last_checked_at": virtool.utils.timestamp(),
        "rid": rid,
        "ready": ready,
        "result": result
    }

    document = await db.analyses.find_one_and_update({"_id": analysis_id, "results.index": sequence_index}, {
        "$set": {
            "results.$.blast": data,
            "updated_at": virtool.utils.timestamp()
        }
    })

    return data, document


async def remove_nuvs_blast(db, analysis_id: str, sequence_index: int):
    """
    Remove the BLAST data for the identified NuVs contig sequence.

    :param db: the application database object
    :param analysis_id: the ID of the analysis containing the sequence
    :param sequence_index: the index of the sequence to remove BLAST data from
    :return:
    """
    await db.analyses.update_one({"_id": analysis_id, "results.index": sequence_index}, {
        "$set": {
            "results.$.blast": None,
            "updated_at": virtool.utils.timestamp()
        }
    })


async def remove_orphaned_directories(app: virtool.types.App):
    """
    Remove all analysis directories for which an analysis document does not exist in the database.

    :param app: the application object

    """
    db = app["db"]

    samples_path = os.path.join(app["settings"]["data_path"], "samples")

    existing_ids = set(await db.analyses.distinct("_id"))

    for sample_id in os.listdir(samples_path):
        analyses_path = os.path.join(samples_path, sample_id, "analysis")

        to_delete = set(os.listdir(analyses_path)) - existing_ids

        for analysis_id in to_delete:
            analysis_path = os.path.join(analyses_path, analysis_id)
            try:
                await app["run_in_thread"](virtool.utils.rm, analysis_path, True)
            except FileNotFoundError:
                pass
