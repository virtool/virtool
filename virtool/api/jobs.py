import os

import virtool.db.jobs
import virtool.http.routes
import virtool.jobs.job
import virtool.jobs.utils
import virtool.resources
import virtool.utils
from virtool.api.utils import compose_regex_query, conflict, json_response, no_content, not_found, paginate

routes = virtool.http.routes.Routes()


@routes.get("/api/jobs")
async def find(req):
    """
    Return a list of job documents.

    """
    db = req.app["db"]

    term = req.query.get("term", None)

    db_query = dict()

    if term:
        db_query.update(compose_regex_query(term, ["task", "user.id"]))

    data = await paginate(
        db.jobs,
        db_query,
        req.query,
        projection=virtool.db.jobs.PROJECTION,
        processor=virtool.db.jobs.processor
    )

    data["documents"].sort(key=lambda d: d["created_at"])

    return json_response(data)


@routes.get("/api/jobs/{job_id}")
async def get(req):
    """
    Return the complete document for a given job.

    """
    job_id = req.match_info["job_id"]

    document = await req.app["db"].jobs.find_one(job_id)

    if not document:
        return not_found()

    return json_response(virtool.utils.base_processor(document))


@routes.put("/api/jobs/{job_id}/cancel", permission="cancel_job")
async def cancel(req):
    """
    Cancel a job.

    """
    db = req.app["db"]

    job_id = req.match_info["job_id"]

    document = await db.jobs.find_one(job_id, ["status"])

    if not document:
        return not_found()

    if not virtool.jobs.utils.is_running_or_waiting(document):
        return conflict("Not cancellable")

    await req.app["jobs"].cancel(job_id)

    document = await db.jobs.find_one(job_id)

    return json_response(virtool.utils.base_processor(document))


@routes.delete("/api/jobs", permission="remove_job")
async def clear(req):
    db = req.app["db"]

    job_filter = req.query.get("filter", None)

    # Remove jobs that completed successfully.
    complete = job_filter in [None, "finished", "complete"]

    # Remove jobs that errored or were cancelled.
    failed = job_filter in [None, "finished", "failed"]

    removed = await virtool.db.jobs.clear(db, complete=complete, failed=failed)

    return json_response({
        "removed": removed
    })


@routes.delete("/api/jobs/{job_id}", permission="remove_job")
async def remove(req):
    """
    Remove a job.

    """
    db = req.app["db"]

    job_id = req.match_info["job_id"]

    document = await db.jobs.find_one(job_id)

    if not document:
        return not_found()

    if virtool.jobs.utils.is_running_or_waiting(document):
        return conflict("Job is running or waiting and cannot be removed")

    # Removed the documents associated with the job ids from the database.
    await db.jobs.delete_one({"_id": job_id})

    try:
        # Calculate the log path and remove the log file. If it exists, return True.
        path = os.path.join(req.app["settings"].get("data_path"), "logs", "jobs", job_id + ".log")
        await virtool.utils.rm(path)
    except OSError:
        pass

    return no_content()


@routes.get("/api/resources")
async def get_resources(req):
    """
    Get a object describing compute resource usage on the server.

    """
    resources = virtool.resources.get()
    req.app["resources"] = resources

    return json_response(resources)
