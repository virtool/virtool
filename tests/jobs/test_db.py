import pytest

import virtool.jobs.db

status = {
    "state": "running",
    "progress": 0.5
}


async def test_processor(dbi, static_time, test_job):
    """
    Test that the dispatch processor properly formats a raw job document into a dispatchable format.

    """
    assert await virtool.jobs.db.processor(dbi, test_job) == {
        "id": "4c530449",
        "created_at": static_time.datetime,
        "args": {
            "workflow": "nuvs",
            "analysis_id": "e410429b",
            "index_id": "465428b0",
            "name": None,
            "sample_id": "1e01a382",
            "username": "igboyes"
        },
        "mem": 16,
        "proc": 10,
        "progress": 1.0,
        "stage": "import_results",
        "state": "complete",
        "task": "build_index",
        "user": {
            "id": "igboyes"
        }
    }


async def test_cancel(dbi, static_time):
    await dbi.jobs.insert_one({
        "_id": "foo",
        "state": "waiting",
        "status": [{
            "state": "running",
            "stage": "foo",
            "error": None,
            "progress": 0.33,
            "timestamp": static_time.datetime
        }]
    })

    await virtool.jobs.db.cancel(dbi, "foo")

    # Check that job document was updated.
    assert await dbi.jobs.find_one() == {
        "_id": "foo",
        "state": "waiting",
        "status": [
            {
                "state": "running",
                "stage": "foo",
                "error": None,
                "progress": 0.33,
                "timestamp": static_time.datetime
            },
            {
                "state": "cancelled",
                "stage": "foo",
                "error": None,
                "progress": 0.33,
                "timestamp": static_time.datetime
            }
        ]
    }


async def test_delete_zombies(dbi):
    documents = [
        {
            "_id": "boo",
            "status": [
                dict(status, state="waiting", progress=0),
                dict(status)
            ]
        },
        {
            "_id": "foo",
            "status": [
                dict(status),
                dict(status, state="cancelled", progress=0.6)
            ]
        },
        {
            "_id": "bar",
            "status": [
                dict(status),
                dict(status, state="error", progress=0.6)
            ]
        },
        {
            "_id": "baz",
            "status": [
                dict(status),
                dict(status, state="complete", progress=1)
            ]
        },
        {
            "_id": "bot",
            "status": [
                dict(status),
                dict(status, state="running", progress=0.3)
            ]
        }
    ]

    await dbi.jobs.insert_many(documents)

    await virtool.jobs.db.delete_zombies(dbi)

    assert await dbi.jobs.find({}, sort=[("_id", 1)]).to_list(None) == [
        documents[2],
        documents[3],
        documents[1]
    ]


@pytest.mark.parametrize("empty", [True, False], ids=["empty", "not-empty"])
async def test_get_waiting_and_running(empty, dbi):
    documents = list()

    if not empty:
        documents.append({
            "_id": "foo",
            "status": [
                {"state": "waiting"},
                {"state": "running"},
            ]
        })

    documents.append({
        "_id": "bar",
        "status": [
            {"state": "waiting"},
            {"state": "running"},
            {"state": "complete"}
        ]
    })

    if not empty:
        documents += [
            {
                "_id": "baz",
                "status": [
                    {"state": "waiting"}
                ]
            },
            {
                "_id": "bat",
                "status": [
                    {"state": "waiting"},
                    {"state": "running"}
                ]
            }
        ]

    await dbi.jobs.insert_many(documents)

    expected = list()

    if not empty:
        expected = ["foo", "baz", "bat"]

    assert await virtool.jobs.db.get_waiting_and_running_ids(dbi) == expected
