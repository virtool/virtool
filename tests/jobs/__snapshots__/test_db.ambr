# serializer version: 1
# name: test_fetch_complete_job
  dict({
    'acquired': False,
    'archived': False,
    'args': dict({
      'analysis_id': 'e410429b',
      'index_id': '465428b0',
      'name': None,
      'sample_id': '1e01a382',
      'username': 'igboyes',
      'workflow': 'nuvs',
    }),
    'created_at': datetime.datetime(2015, 10, 6, 20, 0),
    'id': '4c530449',
    'ping': None,
    'progress': 1,
    'rights': dict({
    }),
    'stage': 'import_results',
    'state': <JobState.COMPLETE: 'complete'>,
    'status': list([
      JobStatus(error=None, progress=0, stage=None, state=<JobState.WAITING: 'waiting'>, step_description=None, step_name=None, timestamp=datetime.datetime(2015, 10, 6, 20, 0)),
      JobStatus(error=None, progress=0, stage=None, state=<JobState.RUNNING: 'running'>, step_description=None, step_name=None, timestamp=datetime.datetime(2015, 10, 6, 20, 0)),
      JobStatus(error=None, progress=0, stage='mk_analysis_dir', state=<JobState.RUNNING: 'running'>, step_description=None, step_name=None, timestamp=datetime.datetime(2015, 10, 6, 20, 0)),
      JobStatus(error=None, progress=1, stage='import_results', state=<JobState.COMPLETE: 'complete'>, step_description=None, step_name=None, timestamp=datetime.datetime(2015, 10, 6, 20, 0)),
    ]),
    'user': UserNested(id='bf1b993c', administrator=False, handle='leeashley'),
    'workflow': 'build_index',
  })
# ---
